modin/utils.py

```diff
@@ -878,3 +878,17 @@ def __init__(self, func: Any):
 
     def __get__(self, instance: Any, owner: Any) -> Any:  # noqa: GL08
         return self.fget(owner)
+
+
+def reload_modin() -> None:
+    """
+    Reload all previously imported Modin modules.
+
+    The call to this function is required
+    if an execution engine has been shut down and
+    is going to be started up once again.
+    """
+    modules = sys.modules.copy()
+    for name, module in modules.items():
+        if name.startswith("modin"):
+            importlib.reload(module)
```

Comment:
Isn't it enough to just do `importlib.reload(modin)`?

Reply:
It's not enough since we have to re-import all previously imported Modin modules.

Reply:
Considering different comments from https://stackoverflow.com/questions/28101895/reloading-packages-and-their-submodules-recursively-in-python, this approach is error prone if you restart the modules not in the correct order.
For example, this comment:
`consider there are two submodules A and B. the objects in A depends on B. Suppose A is reloaded first, A would import the old B (since B has not reloaded). It results in a not fully updated A.`

So the main question here is, can we rely on the order of modules in `sys.modules`?

More reliable would be the engine reinitialization function, which also resets the entire cache that relates to the engine.

Reply:
> So the main question here is, can we rely on the order of modules in sys.modules?

Judging by `print(sys.modules)` the modules are in the order they imported. I think we should reload the modules exactly in this order.

> More reliable would be the engine reinitialization function, which also resets the entire cache that relates to the engine.

IIUC, in that case we will have to manually track each single file related to an engine, what is cumbersome. Also, we may have our own cache(s), which have to be reloaded too.

Having said this, I think we can go with the current approach. If you still think the approach is not reliable, we can reject the changes. Of course, if there are no other ideas ;)

Reply:
> IIUC, in that case we will have to manually track each single file related to an engine, what is cumbersome. Also, we may have our own cache(s), which have to be reloaded too.

I'm considering an observer option, roughly as done [here](https://github.com/modin-project/modin/pull/7307) (instead of reimporting modules). When a situation is detected where ray is restarting, it goes through all known places and updates the cache.

Yes, it is cumbersome, but at least we will know and understand what to expect. When reloading all modules, it is very difficult to predict what may happen. 

On the other hand, I don’t want to be a stopper, so I’ll approve, and then the decision is up to you.